test/functional/multifactor_auths_controller_test.rb

```diff
@@ -39,122 +44,448 @@ class MultifactorAuthsControllerTest < ActionController::TestCase
+        end
+
+        should "not update mfa level" do
+          assert_predicate @user.reload, :mfa_ui_only?
+        end
+
+        should "clear session variables" do
+          assert_nil @controller.session[:mfa_expires_at]
+          assert_nil @controller.session[:level]
+        end
+      end
+
+      context "on PUT to webauthn_update" do
+        setup do
+          put :update, params: { level: "ui_and_api" }
+          @origin = "http://localhost:3000"
+          @rp_id = URI.parse(@origin).host
+          @client = WebAuthn::FakeClient.new(@origin, encoding: false)
+          WebauthnHelpers.create_credential(
+            webauthn_credential: @webauthn_credential,
+            client: @client
+          )
+        end
+
+        context "with correct credentials" do
+          setup do
+            @challenge = session[:webauthn_authentication]["challenge"]
+          end
+
+          context "redirect url is not set" do
             setup do
-              wrong_otp = (ROTP::TOTP.new(@user.mfa_seed).now.to_i.succ % 1_000_000).to_s
-              put :update, params: { otp: wrong_otp, level: "disabled" }
+              put(
+                :webauthn_update,
+                params: {
+                  credentials:
+                    WebauthnHelpers.get_result(
+                      client: @client,
+                      challenge: @challenge
+                    )
+                }
+              )
             end
 
-            should respond_with :redirect
             should redirect_to("the settings page") { edit_settings_path }
-            should set_flash.to("Your OTP code is incorrect.")
 
-            should "keep mfa enabled" do
-              assert_predicate @user.reload, :mfa_enabled?
+            should "update mfa level" do
+              assert_predicate @user.reload, :mfa_ui_and_api?
             end
 
-            should "not send mfa disabled email" do
-              assert_emails 0
+            should "clear session variables" do
+              assert_nil @controller.session[:mfa_expires_at]
+              assert_nil @controller.session[:level]
+              assert_nil @controller.session[:webauthn_authentication]
             end
           end
+
+          context "when redirect url is set" do
+            setup do
+              @controller.session["mfa_redirect_uri"] = profile_api_keys_path
+              put(
+                :webauthn_update,
+                params: {
+                  credentials:
+                    WebauthnHelpers.get_result(
+                      client: @client,
+                      challenge: @challenge
+                    )
+                }
+              )
+            end
+
+            should redirect_to("the api keys index") { profile_api_keys_path }
+          end
         end
 
-        context "on updating to ui_only, flash banner is set and mfa level is unchanged" do
```

Comment:
TODO: missing test, update `ui_only` should fail.

Reply:
Can you clarify what you mean? Do you mean we still need to test updating `ui_only`, but assert that it is not possible/it fails? As a follow up, is there any reason that it is still part of the enum?

Reply:
Eric and I discussed this. For others,
> Do you mean we still need to test updating ui_only, but assert that it is not possible/it fails? 
Yes, we should make sure that updating to `ui_only` is possible

> As a follow up, is there any reason that it is still part of the enum?
No reason, just haven't found the time to clean this up yet.